include/zenoh-pico/collections/refcount.h

```diff
@@ -38,122 +41,292 @@
+#define _ZP_RC_OP_CHECK_WEAK_CNT _z_atomic_compare_exchange_strong(&p->in->_weak_cnt, &cmp_val, cmp_val)
 #define _ZP_RC_OP_SYNC atomic_thread_fence(_z_memory_order_acquire);
+#define _ZP_RC_OP_UPGRADE_CAS_LOOP                                                                              \
+    unsigned int prev = _z_atomic_load_explicit(&p->in->_strong_cnt, _z_memory_order_relaxed);                  \
+    while ((prev != 0) && (prev < UINT32_MAX)) {                                                                \
+        unsigned int next = prev + 1;                                                                           \
+        if (_z_atomic_compare_exchange_weak_explicit(&p->in->_strong_cnt, &prev, next, _z_memory_order_acquire, \
+                                                     _z_memory_order_relaxed)) {                                \
+            _ZP_RC_OP_INCR_WEAK_CNT                                                                             \
+            c.in = p->in;                                                                                       \
+            return c;                                                                                           \
+        } else {                                                                                                \
+            prev = _z_atomic_load_explicit(&p->in->_strong_cnt, _z_memory_order_relaxed);                       \
+        }                                                                                                       \
+    }
 
 #else  // ZENOH_C_STANDARD == 99
 #ifdef ZENOH_COMPILER_GCC
 
 // c99 gcc sync builtin variant
 #define _ZP_RC_CNT_TYPE unsigned int
-#define _ZP_RC_OP_INIT_CNT                              \
-    __sync_fetch_and_and(&p.in->_cnt, (unsigned int)0); \
-    __sync_fetch_and_add(&p.in->_cnt, (unsigned int)1);
-#define _ZP_RC_OP_INCR_CNT __sync_fetch_and_add(&p->in->_cnt, (unsigned int)1);
-#define _ZP_RC_OP_DECR_AND_CMP __sync_fetch_and_sub(&p->in->_cnt, (unsigned int)1) > (unsigned int)1
+#define _ZP_RC_OP_INIT_CNT                                     \
+    __sync_fetch_and_and(&p.in->_strong_cnt, (unsigned int)0); \
+    __sync_fetch_and_add(&p.in->_strong_cnt, (unsigned int)1); \
+    __sync_fetch_and_and(&p.in->_weak_cnt, (unsigned int)0);   \
+    __sync_fetch_and_add(&p.in->_weak_cnt, (unsigned int)1);
+#define _ZP_RC_OP_INCR_STRONG_CNT __sync_fetch_and_add(&p->in->_strong_cnt, (unsigned int)1);
+#define _ZP_RC_OP_INCR_WEAK_CNT __sync_fetch_and_add(&p->in->_weak_cnt, (unsigned int)1);
+#define _ZP_RC_OP_DECR_AND_CMP_STRONG __sync_fetch_and_sub(&p->in->_strong_cnt, (unsigned int)1) > (unsigned int)1
+#define _ZP_RC_OP_DECR_AND_CMP_WEAK __sync_fetch_and_sub(&p->in->_weak_cnt, (unsigned int)1) > (unsigned int)1
+#define _ZP_RC_OP_CHECK_STRONG_CNT __sync_bool_compare_and_swap(&p->in->_strong_cnt, cmp_val, cmp_val)
```

Comment:
maybe it would be better to allow macros to accept cmp_val as an argument ? It would likely increase code readability.